### DISCUSSION TITLE
Tools that support the DTCG format

### DISCUSSION THREAD

**Discussion by c1rrus** [design-tokens/community-group#312] [Show and tell]
Do you maintain a tool (or anything else - e.g. a software library) that supports reading and/or writing DTCG files? Then please add a link and some details as a comment here in the following format:

```markdown
### [Name of your tool](https://website.of/your/tool)

![Logo or screenshot of your tool](https://link.to/the/image)

Brief description of your tool (1-2 paragraphs). E.g. What does it do? Who's it for?

#### Supported spec version(s)

<!-- Add or delete as appropriate -->
* [Live draft](https://tr.designtokens.org/format/)
* [2nd editor's draft](https://second-editors-draft.tr.designtokens.org/format/)
* [1st editor's draft](https://first-editors-draft.tr.designtokens.org/format/)
```

🙏 **Please only add things here that have already shipped support for the DTCG file format.** If you are planning to support the format but haven't released it yet, that's awesome. However, please wait until you actually ship support before promoting your tool here.

♻️ If details change (e.g. you add more features to your tool), please edit your existing comment rather than adding a new one.

⚠️ The DTCG maintainers reserve the right to remove spammy or inappropriate comments.

### Motivation

This list will hopefully be useful for:

* Letting tool makers promote their good work and get some well-deserved kudos 😄 
* Providing the community with a list of tools they can use to play with DTCG files _right now_
* Giving the DTCG format editors visibility of how folks are using the format

**Comment by andrico1234** [comment]
### [Design Token Validator](https://animaapp.github.io/design-token-validator-site/)

Check that your design tokens adhere to the DTCG spec with this free open source _Design Tokens Validator_ 

To get started, paste the JSON for your design tokens, or upload a file directly. The site will then validate your tokens and present a list of helpful messages, which link to relevant parts of the spec, to help fix any problems.

![image](https://github.com/design-tokens/community-group/assets/17087167/17b8c446-a340-4138-bb17-23d7b6a06379)

#### Supported spec version(s)
- [Live draft](https://tr.designtokens.org/format/)

**Comment by drwpow** [comment]
### [Cobalt CLI](https://cobalt-ui.pages.dev)

<a href="https://cobalt-ui.pages.dev" target="_blank" rel="noopener noreferrer"><img src="https://github.com/design-tokens/community-group/assets/1369770/e553aeed-5da4-4ec0-ac8e-b7ae6257ac97" /></a>

Cobalt is [an open-source, MIT-licensed](https://github.com/drwpow/cobalt-ui) frontend toolchain that:

- can validate any DTCG tokens schema and give detailed errors
- has existing integrations for [JavaScript](https://cobalt-ui.pages.dev/integrations/js), [TypeScript](https://cobalt-ui.pages.dev/integrations/js), [CSS](https://cobalt-ui.pages.dev/integrations/css)/[Sass](https://cobalt-ui.pages.dev/integrations/sass), [JSON](https://cobalt-ui.pages.dev/integrations/json) and [Tailwind CSS](https://cobalt-ui.pages.dev/integrations/tailwind)
- Used in production for several sites ([Snyk](https://snyk.io/) and [Tidepool](https://www.tidepool.so/))
- [has a plugin system](https://cobalt-ui.pages.dev/guides/getting-started) that can support any build target
- can bundle multiple tokens schemas together
- supports the [modes extension](https://cobalt-ui.pages.dev/guides/modes) for light & dark mode

#### Supported spec version(s)

* [Live draft](https://tr.designtokens.org/format/)

**Comment by PavelLaptev** [comment]
## [TokensBrücke — Figma plugin](https://www.figma.com/community/plugin/1254538877056388290)

<a href="https://www.figma.com/community/plugin/1254538877056388290" target="_blank">
<img src="https://raw.githubusercontent.com/tokens-bruecke/figma-plugin/main/readme-assets/preview.webp" alt="preview" width="100%">
</a>

TokensBrücke is a Figma plugin that simplifies the process of converting Figma variables into design-tokens JSON. The plugin generates JSON files compatible with the latest DTCG specification.

- [Github repo](https://github.com/tokens-bruecke/figma-plugin)
- [Style Dictionary utils](https://github.com/tokens-bruecke/sd-utils)

#### Supported spec version(s)
- [Live draft](https://tr.designtokens.org/format/)

**Comment by kaelig** [comment]
@universse 👋🏻 It'd be great to see https://github.com/universse/theminglayer showcased here!

**Comment by tmdvs** [comment]
### [Sketch](https://www.sketch.com/handoff/)

![Design Tokes in the Sketch web app](https://github.com/design-tokens/community-group/assets/177527/27708ba9-c91e-459a-9ef5-76325e233aeb)

Sketch is Mac app for designers to create, team up, [prototype](https://www.sketch.com/prototype/), as well as a web app for everyone else to browse, give feedback, inspect, and [handoff](https://www.sketch.com/handoff/) — in any browser. A complete design platform, made by a sustainable indie company since 2010. 

You can export your Color Variables, Layer Styles, and Text Styles as Design Tokens from the web app. You can download your design tokens or create an always-up-to-date public link for them. ([Read the docs](https://www.sketch.com/docs/developer-handoff/#design-token-exports))

#### Supported spec version(s)
* [Live draft](https://tr.designtokens.org/format/)

**Comment by nclsndr** [comment]
### [Design Tokens Format Module definitions](https://github.com/nclsndr/design-tokens-format-module)

A npm package providing implementation agnostic JavaScript/TypeScript definitions from the Design Tokens Format Module specification, for library developers and tooling creators.

In short, it exposes constants (fontWeights, strokeStyleLineButt...) and TS types to avoid repeating them in every projects.

#### Supported spec version(s)


* [Live draft](https://tr.designtokens.org/format/)

**Comment by bartveneman** [comment]
### [Project Wallace](https://www.projectwallace.com/design-tokens)

![Image](https://github.com/user-attachments/assets/e07e2eca-5aaa-461e-a57a-2e66f8c20199)

Generate design tokens from CSS. Because Project Wallace [analyzes](https://github.com/projectwallace/css-analyzer) your CSS in depth already, we can convert the analysis into design tokens.

#### Supported spec version(s)


* [Live draft](https://tr.designtokens.org/format/)

**Reply by bartveneman** [reply]
The logic that converts CSS into Design Tokens is now a separate library: https://github.com/projectwallace/css-design-tokens/

**Comment by phuoccss** [comment]
### [W3C Design Token Standard Schema](https://github.com/universse/w3c-design-tokens-standard-schema)

Provides runtime validation and type inference for W3C design tokens with [Standard Schema](https://standardschema.dev).

#### Supported spec version(s)

* [Live draft](https://tr.designtokens.org/format/)

**Comment by artursopelnik** [comment]
### [Design Token Validator](https://designtoken-validator.sotec-solutions.com)

![Design Token Validator ](https://github.com/user-attachments/assets/747018ba-0d2c-4ccf-872b-7710771b34d6)

**Validate your design tokens against the W3C DTCG spec to ensure compatibility.**  
Paste JSON/JSON5 or upload a file. The tool checks your tokens and gives clear, actionable feedback. A CLI version for workflow integration is coming soon.

#### Supported spec version(s)

* [Live draft](https://tr.designtokens.org/format/)

**Comment by bennypowers** [comment]
### [Design Tokens Language Server](https://github.com/bennypowers/design-tokens-language-server)

> [!IMPORTANT]
> Archived and moved to [asimonim](https://github.com/design-tokens/community-group/discussions/312#discussioncomment-16179760)

Editor tooling for DTCG design tokens. It's got **Hover**, **Go to Definition**, **Find References**, **Autocomplete**, and more for your favourite editors like neovim, vscode, or zed. Supports tokens defined in DTCG format JSON or YAML files, and CSS files.

![Image](https://github.com/user-attachments/assets/d1e92fa6-b82d-43b8-85a5-cdcb5323842c)
![Image](https://github.com/user-attachments/assets/1a769da6-3f0c-4ceb-b8bb-e0e083637cb9)

#### Supported spec version(s)
* [Live draft](https://tr.designtokens.org/format/)

**Comment by cmjio** [comment]
### [Design Tokens Language](https://designtokenslang.com)

![screenshot of language documentation site](https://i.postimg.cc/sDyM3dNS/screenshot.png)


DTL (Design Token Language) is a soon to be open-source specification and syntax designed to bring type safety, logic, and structure to design system management. It serves as a unified, platform-agnostic source of truth that allows teams to define styles once and automatically compile them into usable code for CSS, iOS, Android, and Flutter. By replacing scattered JSON files with a robust language ecosystem, DTL helps developers and designers automate their workflows and maintain consistency at scale.

#### Supported spec version(s)


* [Live draft](https://tr.designtokens.org/format/)

**Comment by thevangelist** [comment]
I attempted to create a schema validator for the most recent spec. https://dembrandt.github.io/dtcg-validator

<img width="1703" height="1199" alt="Screenshot 2025-12-29 at 20 22 43" src="https://github.com/user-attachments/assets/ce39eb18-52ed-47a8-afab-3e255415519c" />

**Reply by thevangelist** [reply]
https://www.dembrandt.com/validator updated address

**Comment by TrySound** [comment]
# Engramma

Recently started building design tokens apps

https://engramma.dev
https://github.com/TrySound/engramma

<img width="1508" height="742" alt="image" src="https://github.com/user-attachments/assets/a638b916-ff20-4393-ab6b-4c024328e5ca" />

**Reply by artursopelnik** [reply]
Nice! Works great!

**Comment by timges** [comment]
# [Dispersa](https://dispersa.dev)

Translation lib for design tokens, similar to Terrazzo and Style Dictionary.

<img width="1437" height="1148" alt="image" src="https://github.com/user-attachments/assets/b1cc4384-2780-4a92-be7f-bed0a9b75767" />

## What it does

- **2025.10 spec alignment**, including token *and* resolver validation via JSON Schemas.
- Built-in output targets: **CSS, JSON, JS, Tailwind, Android, iOS** with some customizable options  
- TypeScript Type generation
- A wide range of built-in transforms and filters (color spaces, dimensions, naming, etc.)  
- Fully extensible pipeline:
  - Custom transforms  
  - Custom filters  
  - Custom preprocessors  
  - Completely custom output targets  
- lifecycle hooks (onBuildStart, onBuildEnd)
- CLI + Node API  

---

One might be wondering:  
*Why build this when Terrazzo and Style Dictionary already exist?*

Fair question 😄

When I started the project some time ago, none of the established tools were implementing the stable 2025.10 spec yet. Resolver support especially didn’t seem close to landing. So I just gave it a shot. Terrazzo landed the features eventually, but I still wanted to finish what I started

And well… here we are. It's Open Source (MIT), the repo can be fond at [https://github.com/dispersa-core/dispersa](https://github.com/dispersa-core/dispersa), while the docs live at [https://dispersa.dev](https://dispersa.dev)

I'm neither a mobile dev, nor do I work with tailwind a lot, which is why those ouput targets may still be a bit off. If you're into that, or find anything else, feel free to contribute :)


#### Supported Spec Versions
* [2025.10](https://www.designtokens.org/tr/2025.10/)

**Reply by drwpow** [reply]
> Why build this when Terrazzo and Style Dictionary already exist?

Glad you built it! There’s no such thing as a tool for everyone. 🙂  Users only benefit from more options in OSS. And usually every attempt at a problem yields new ideas that wouldn’t have come about otherwise. Eager to check this out!

**Reply by timges** [reply]
> Users only benefit from more options in OSS. And usually every attempt at a problem yields new ideas that wouldn’t have come about otherwise. 

I agree! Have fun playing around 🙂🤙🏼

**Comment by bennypowers** [comment]
### [asimonim][asimonim]

![asimonim - israeli phone tokens](https://upload.wikimedia.org/wikipedia/commons/1/17/%D7%90%D7%A1%D7%99%D7%9E%D7%95%D7%9F_%D7%99%D7%A9%D7%A8%D7%90%D7%9C%D7%99.jpg)

> pr. Ahh-see-moh-NEEM (אסימונים). Israeli Phone tokens in use up until the 90s

The design tokens multitool. Convert DTCG tokens to CSS, SCSS, JavaScript/TypeScript, Swift, Android XML, and more. Validate, search, and explore token files from the command line. Get completions, hover, and diagnostics in your editor.

asimonim is the successor to [design-tokens-language-server][dtls]. All DTLS features (completions, hover, go-to-definition, diagnostics, color previews, etc.) are included, along with a CLI for token validation, search, and multi-format conversion. Existing VS Code and Zed extension users will receive updates automatically.

#### Convert

Generate platform-native output from your DTCG token files:

- **CSS** custom properties (`:root`/`:host` selector, Lit `css` tagged template wrapper)
- **SCSS** variables
- **JavaScript/TypeScript** (ESM/CJS, `as const` or JSDoc types, individual exports or `TokenMap` class)
- **Swift** constants with native `SwiftUI.Color` and color space mapping
- **Android** XML resources with automatic wide-gamut to sRGB hex downsampling
- **Editor snippets** for VS Code, TextMate, and Zed
- **Schema conversion** between Draft and 2025.10 (string colors to structured colors, curly-brace refs to `$ref`, etc.)
- DTCG JSON roundtrip and flat key-value JSON

#### CLI

- **Validate** token files for schema compliance, circular references, and missing aliases
- **List** tokens with filtering by type, group, or deprecation status
- **Search** by name, value, or type with full regex support
- **Resolver documents** for composing multi-file token systems
- **Auto-discovery** of tokens from npm dependencies via `designTokens` field or export conditions

#### Language Server

Built-in LSP (`asimonim lsp`) for CSS, HTML, JavaScript, TypeScript, JSON, and YAML:

- **Completions** for token references as CSS custom properties or curly-brace aliases
- **Hover** with resolved values, type info, and resolution chains
- **Go to Definition** and **Find References** across files
- **Diagnostics** for invalid references, circular aliases, and schema violations
- **Document Colors** with inline previews and color picker
- **Semantic Tokens** and **Code Actions**

#### Editor Extensions

- [VS Code][vscode] (marketplace)
- [Zed][zed] (extension registry)
- [Claude Code][claude] (plugin)
- Neovim, Helix, and any editor with LSP support (via `asimonim lsp`)

#### Install

🐋 NPM:
```bash
npm install -D @pwrs/asimonim    # npm
```
🐿️ Go:
```bash
go install bennypowers.dev/asimonim@latest  # go
```
🐧 Gentoo:
```shell
# eselect repository enable bennypowers
# emerge dev-util/asimonim
```

#### Supported spec version(s)

* [Live draft][draft]
* [Stable (2025.10)][stable] -- including structured colors, `$ref`, `$extends`, resolver documents

[asimonim]: https://github.com/bennypowers/asimonim
[dtls]: https://github.com/bennypowers/design-tokens-language-server
[vscode]: https://marketplace.visualstudio.com/items?itemName=pwrs.design-tokens-language-server-vscode
[zed]: https://github.com/zed-industries/extensions/tree/main/extensions/design-tokens
[claude]: https://github.com/bennypowers/asimonim/tree/main/.claude-plugin
[draft]: https://tr.designtokens.org/format/
[stable]: https://www.designtokens.org/schemas/2025.10.json

**Comment by dbychinsky** [comment]
## Design Token Kit

**Design Token Kit** is an open-source toolkit for working with design tokens.

Website:
https://design-token-kit.github.io/

The project uses DTCG JSON as its primary format and internal processing model.
It also supports HRDT YAML as a more human-readable authoring format, and
DESIGN.md as a Markdown-based documentation format with YAML frontmatter,
inspired by the Google Labs DESIGN.md approach. Both HRDT YAML and DESIGN.md
sources are normalized into DTCG JSON before validation, conversion, and
artifact generation. The toolkit can also export supported DTCG token trees
back to DESIGN.md.

The project consists of the core library `@design-token-kit/core`, the
`dtokens` CLI available through the `@design-token-kit/cli` package, and a
Figma plugin for exporting design tokens from Figma files.

Main features:

* validating tokens against the DTCG schema (DTCG 2025.10);
* checking token model correctness, including unresolved references, circular
  references, reference-to-group errors, type mismatches, duplicate gradient
  stops, and deprecated token usage warnings;
* lint checks and diagnostics for token structure issues, including root-layer
  naming, cross-layer references, raw value placement, empty groups, and
  missing descriptions;
* running checks at different levels: `schema`, `validate`, and `lint`;
* converting between supported formats;
* generating CSS custom properties;
* generating SCSS variables;
* generating Tailwind CSS v4 `@theme` output;
* supporting theme-aware conversion workflows, including multi-theme SCSS
  output;
* generating SwiftUI source code with support for base and theme token sets
  using nested enums or an additional `Theme` struct layer;
* generating Android source code with support for base and theme token sets;
* generating static HTML showcase pages;
* collecting token statistics as text or HTML reports;
* exporting Figma Variables and local styles to DTCG JSON;
* preserving Figma variable aliases as DTCG references when possible;
* exporting Figma variable modes as separate theme token files;
* converting exported tokens to CSS, SCSS, Tailwind CSS v4, Android XML, and
  SwiftUI;
* displaying token counts, skipped entries, and warnings in the Figma plugin.

**Supported formats:**

* DTCG JSON
* HRDT YAML
* DESIGN.md

**CLI interface:** available via the `dtokens` command.

**Figma plugin:** exports design tokens from Figma Variables and local styles
through the Design Token Kit pipeline.

Figma Community:
https://www.figma.com/community/plugin/1661669063434861129

**GitHub repository:**
https://github.com/design-token-kit/design-token-kit

**npm packages:**
CLI: https://www.npmjs.com/package/@design-token-kit/cli
Core: https://www.npmjs.com/package/@design-token-kit/core

<table>
  <tr>
<td>
<img width="1891" height="891" alt="Screenshot 2026-07-29 124555" src="https://github.com/user-attachments/assets/878e93af-d39a-4d5f-b63b-8c45751293ed" />

</td>
    <td>
<img width="1899" height="908" alt="showcase" src="https://github.com/user-attachments/assets/e8f55d08-7eff-48cc-907b-b2f71124fa9e" />
    </td>
    <td>
<img width="1901" height="910" alt="statistic" src="https://github.com/user-attachments/assets/53ff256b-b1cc-493f-9528-2f37cafb9be8" />
    </td>
  </tr>
</table>

**Supported DTCG specification version:**
DTCG 2025.10

**Comment by LE-VAI** [comment]
## [@designesy/tokens](https://www.npmjs.com/package/@designesy/tokens)

A zero-dependency CLI that validates design token JSON files against the [DTCG 2025.10 Final Community Group Report](https://www.designtokens.org/TR/2025.10/) — the spec's first stable version, published October 28, 2025.

Runs 10 conformance checks (t01–t10) covering structural correctness: `$type`/`$value`/`$description` presence, structured colors (colorSpace + components rather than bare hex), valid `$schema` pointer to designtokens.org, correct dimension units, custom type namespacing, and type drift detection. Returns PASS/FAIL/WARN per check with a numeric score and validity verdict.

Designed as a lightweight lint step for CI pipelines — install with `npm i -D @designesy/tokens`, run with `npx designesy-tokens tokens.json`, or use it as a library import. The full source is MIT-licensed and the package has zero runtime dependencies.

#### Supported spec version(s)
- [2025.10 Final Community Group Report](https://www.designtokens.org/TR/2025.10/)

**Comment by julien-deramond** [comment]
Hi everyone 👋

A few thousand design tokens isn't a lot of data, but the references between them are hard to hold in your head: one primitive feeds a semantic token, which feeds a component token, sometimes across a dozen files, and you rebuild that chain from memory every time you touch one.

I wanted a new representation of them, so I've been building [**dtgraph**](https://github.com/julien-deramond/dtgraph), the latest of a few experiments: it parses DTCG files and draws the tokens as a node graph, one node per token and one edge per reference.

[![The dtgraph playground: a design token graph clustered by group, with primitives sized by how many tokens reference them](https://raw.githubusercontent.com/julien-deramond/dtgraph/6b0d79a/media/playground.png)](https://julien-deramond.github.io/dtgraph/)

Playground (client-side, nothing is uploaded): https://julien-deramond.github.io/dtgraph/
Repository: https://github.com/julien-deramond/dtgraph
Docs: https://julien-deramond.github.io/dtgraph/docs/

**Comment by hipuku** [comment]
### [tokenise](https://tokenise.hipuku.dev)

![tokenise](https://tokenise.hipuku.dev/og.png)

A browser tool that converts design tokens between DTCG JSON, Figma's variables export, Tokens Studio JSON and a Tailwind v4 `@theme` stylesheet, in any direction. Every conversion comes with a report of each token that did not come through exactly, and why, grouped as dropped, changed or same meaning. It is for design system teams moving tokens between tools who want to know what a conversion loses before it ships. Free, no sign-up, open source (MIT): https://github.com/hipuku/tokenise

On the DTCG side it reads single 2025.10 token files, resolver documents with their sets and modifiers, and the pre-2025.10 shape where colours and dimensions are CSS strings, and it reports anything it had to read leniently. It writes 2025.10, with modes as a resolver document. Every format goes through one internal model shaped like DTCG 2025.10, so there are four readers and four writers rather than twelve pairwise conversions.

#### Supported spec version(s)

* [2025.10](https://www.designtokens.org/TR/2025.10/) (read and write, including resolver documents)
* Pre-2025.10 drafts (read)

**Comment by zetareticoli** [comment]
### [TokenLens](https://tokenlens.app)

<img width="1440" height="1024" alt="tokenlens-app-website" src="https://github.com/user-attachments/assets/337bd1bd-4f6c-4a2b-a0bd-735665ffd2cb" />


Analyze design token usage in CSS file and DTCG format for JSON files.
